Human readable representation of link and chain

wkentaro opened the pull request:
## Why?

Currently, the representation of models implemented using Chain and ChainList is not so helpful to get to know about the model structure as follows.

```
% cat model_resnet.py
from chainer.links.model.vision.resnet import ResNet50Layers

model = ResNet50Layers(pretrained_model=None)
print(model)

% python model_resnet.py
<chainer.links.model.vision.resnet.ResNet50Layers object at 0x108fc4e48>
```

I think it is useful for users if registered links are printed, even though chainer has the define-by-run policy.

Below demonstrates the representation of ResNet50 with the pathces on this PR, and it seems better than before to understand its architecture.

```
ResNet50Layers(
  (bn1): BatchNormalization(size=64, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
  (conv1): Convolution2D(in_channels=3, out_channels=64, ksize=7, stride=2, pad=3, nobias=False, dilate=1, groups=1),
  (fc6): Linear(in_size=2048, out_size=1000, nobias=False),
  (res2): BuildingBlock(
    (a): BottleneckA(
      (bn1): BatchNormalization(size=64, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn2): BatchNormalization(size=64, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn3): BatchNormalization(size=256, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn4): BatchNormalization(size=256, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (conv1): Convolution2D(in_channels=64, out_channels=64, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
      (conv2): Convolution2D(in_channels=64, out_channels=64, ksize=3, stride=1, pad=1, nobias=True, dilate=1, groups=1),
      (conv3): Convolution2D(in_channels=64, out_channels=256, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
      (conv4): Convolution2D(in_channels=64, out_channels=256, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
    ),
    (b1): BottleneckB(
      (bn1): BatchNormalization(size=64, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn2): BatchNormalization(size=64, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn3): BatchNormalization(size=256, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (conv1): Convolution2D(in_channels=256, out_channels=64, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
      (conv2): Convolution2D(in_channels=64, out_channels=64, ksize=3, stride=1, pad=1, nobias=True, dilate=1, groups=1),
      (conv3): Convolution2D(in_channels=64, out_channels=256, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
    ),
    (b2): BottleneckB(
      (bn1): BatchNormalization(size=64, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn2): BatchNormalization(size=64, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn3): BatchNormalization(size=256, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (conv1): Convolution2D(in_channels=256, out_channels=64, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
      (conv2): Convolution2D(in_channels=64, out_channels=64, ksize=3, stride=1, pad=1, nobias=True, dilate=1, groups=1),
      (conv3): Convolution2D(in_channels=64, out_channels=256, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
    ),
  ),
  (res3): BuildingBlock(
    (a): BottleneckA(
      (bn1): BatchNormalization(size=128, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn2): BatchNormalization(size=128, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn3): BatchNormalization(size=512, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn4): BatchNormalization(size=512, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (conv1): Convolution2D(in_channels=256, out_channels=128, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
      (conv2): Convolution2D(in_channels=128, out_channels=128, ksize=3, stride=2, pad=1, nobias=True, dilate=1, groups=1),
      (conv3): Convolution2D(in_channels=128, out_channels=512, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
      (conv4): Convolution2D(in_channels=256, out_channels=512, ksize=1, stride=2, pad=0, nobias=True, dilate=1, groups=1),
    ),
    (b1): BottleneckB(
      (bn1): BatchNormalization(size=128, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn2): BatchNormalization(size=128, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn3): BatchNormalization(size=512, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (conv1): Convolution2D(in_channels=512, out_channels=128, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
      (conv2): Convolution2D(in_channels=128, out_channels=128, ksize=3, stride=1, pad=1, nobias=True, dilate=1, groups=1),
      (conv3): Convolution2D(in_channels=128, out_channels=512, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
    ),
    (b2): BottleneckB(
      (bn1): BatchNormalization(size=128, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn2): BatchNormalization(size=128, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn3): BatchNormalization(size=512, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (conv1): Convolution2D(in_channels=512, out_channels=128, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
      (conv2): Convolution2D(in_channels=128, out_channels=128, ksize=3, stride=1, pad=1, nobias=True, dilate=1, groups=1),
      (conv3): Convolution2D(in_channels=128, out_channels=512, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
    ),
    (b3): BottleneckB(
      (bn1): BatchNormalization(size=128, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn2): BatchNormalization(size=128, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn3): BatchNormalization(size=512, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (conv1): Convolution2D(in_channels=512, out_channels=128, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
      (conv2): Convolution2D(in_channels=128, out_channels=128, ksize=3, stride=1, pad=1, nobias=True, dilate=1, groups=1),
      (conv3): Convolution2D(in_channels=128, out_channels=512, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
    ),
  ),
  (res4): BuildingBlock(
    (a): BottleneckA(
      (bn1): BatchNormalization(size=256, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn2): BatchNormalization(size=256, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn3): BatchNormalization(size=1024, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn4): BatchNormalization(size=1024, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (conv1): Convolution2D(in_channels=512, out_channels=256, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
      (conv2): Convolution2D(in_channels=256, out_channels=256, ksize=3, stride=2, pad=1, nobias=True, dilate=1, groups=1),
      (conv3): Convolution2D(in_channels=256, out_channels=1024, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
      (conv4): Convolution2D(in_channels=512, out_channels=1024, ksize=1, stride=2, pad=0, nobias=True, dilate=1, groups=1),
    ),
    (b1): BottleneckB(
      (bn1): BatchNormalization(size=256, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn2): BatchNormalization(size=256, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn3): BatchNormalization(size=1024, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (conv1): Convolution2D(in_channels=1024, out_channels=256, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
      (conv2): Convolution2D(in_channels=256, out_channels=256, ksize=3, stride=1, pad=1, nobias=True, dilate=1, groups=1),
      (conv3): Convolution2D(in_channels=256, out_channels=1024, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
    ),
    (b2): BottleneckB(
      (bn1): BatchNormalization(size=256, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn2): BatchNormalization(size=256, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn3): BatchNormalization(size=1024, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (conv1): Convolution2D(in_channels=1024, out_channels=256, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
      (conv2): Convolution2D(in_channels=256, out_channels=256, ksize=3, stride=1, pad=1, nobias=True, dilate=1, groups=1),
      (conv3): Convolution2D(in_channels=256, out_channels=1024, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
    ),
    (b3): BottleneckB(
      (bn1): BatchNormalization(size=256, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn2): BatchNormalization(size=256, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn3): BatchNormalization(size=1024, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (conv1): Convolution2D(in_channels=1024, out_channels=256, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
      (conv2): Convolution2D(in_channels=256, out_channels=256, ksize=3, stride=1, pad=1, nobias=True, dilate=1, groups=1),
      (conv3): Convolution2D(in_channels=256, out_channels=1024, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
    ),
    (b4): BottleneckB(
      (bn1): BatchNormalization(size=256, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn2): BatchNormalization(size=256, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn3): BatchNormalization(size=1024, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (conv1): Convolution2D(in_channels=1024, out_channels=256, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
      (conv2): Convolution2D(in_channels=256, out_channels=256, ksize=3, stride=1, pad=1, nobias=True, dilate=1, groups=1),
      (conv3): Convolution2D(in_channels=256, out_channels=1024, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
    ),
    (b5): BottleneckB(
      (bn1): BatchNormalization(size=256, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn2): BatchNormalization(size=256, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn3): BatchNormalization(size=1024, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (conv1): Convolution2D(in_channels=1024, out_channels=256, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
      (conv2): Convolution2D(in_channels=256, out_channels=256, ksize=3, stride=1, pad=1, nobias=True, dilate=1, groups=1),
      (conv3): Convolution2D(in_channels=256, out_channels=1024, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
    ),
  ),
  (res5): BuildingBlock(
    (a): BottleneckA(
      (bn1): BatchNormalization(size=512, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn2): BatchNormalization(size=512, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn3): BatchNormalization(size=2048, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn4): BatchNormalization(size=2048, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (conv1): Convolution2D(in_channels=1024, out_channels=512, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
      (conv2): Convolution2D(in_channels=512, out_channels=512, ksize=3, stride=2, pad=1, nobias=True, dilate=1, groups=1),
      (conv3): Convolution2D(in_channels=512, out_channels=2048, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
      (conv4): Convolution2D(in_channels=1024, out_channels=2048, ksize=1, stride=2, pad=0, nobias=True, dilate=1, groups=1),
    ),
    (b1): BottleneckB(
      (bn1): BatchNormalization(size=512, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn2): BatchNormalization(size=512, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn3): BatchNormalization(size=2048, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (conv1): Convolution2D(in_channels=2048, out_channels=512, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
      (conv2): Convolution2D(in_channels=512, out_channels=512, ksize=3, stride=1, pad=1, nobias=True, dilate=1, groups=1),
      (conv3): Convolution2D(in_channels=512, out_channels=2048, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
    ),
    (b2): BottleneckB(
      (bn1): BatchNormalization(size=512, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn2): BatchNormalization(size=512, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (bn3): BatchNormalization(size=2048, decay=0.9, eps=2e-05, dtype=<class 'numpy.float32'>, use_gamma=True, use_beta=True),
      (conv1): Convolution2D(in_channels=2048, out_channels=512, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
      (conv2): Convolution2D(in_channels=512, out_channels=512, ksize=3, stride=1, pad=1, nobias=True, dilate=1, groups=1),
      (conv3): Convolution2D(in_channels=512, out_channels=2048, ksize=1, stride=1, pad=0, nobias=True, dilate=1, groups=1),
    ),
  ),
)
```

## Solution

Implement `__repr__` of `Link`, `Chain` and `ChainList`.
For `Link`, I add a new attribute `self._kwargs` which should be overwritten by child classes to represent the arguments for `__init__ `.
For example,  in `Linear` link, it adds `in_size`, `out_size`, and `nobias` to the `self._kwargs`.

kmaehashi commented:
Thank you for the PR. The overall idea looks good to me!
General comments:

* Could you use `__str__` instead of `__repr__`? (c.f., https://docs.python.org/3/reference/datamodel.html#object.__repr__ )
* I think it is better to expose `self._kwargs` (or function to register such keyword arguments) for so that user-defined Links can use it.

wkentaro commented:
> Could you use __str__ instead of __repr__?

Do you have an idea of which of them we use?
I referred this link: https://stackoverflow.com/questions/1436703/difference-between-str-and-repr for the general idea.
And I agreed with the below conclusion of the post and ended up with the implementation of this PR.

>> Implement __repr__ for any class you implement. This should be second nature. Implement __str__ if you think it would be useful to have a string version which errs on the side of more readability in favor of more ambiguity.

-----

> I think it is better to expose self._kwargs (or function to register such keyword arguments) for so that user-defined Links can use it.

Could you give me an example of that kind of use cases?
I am conservative to create new public attribute since it may restrict future development to keep the backward compatibility.

wkentaro commented:
I also added a representation for chainer.Sequential

```
% python model_sequential.py
Chain(
  (seq1): Sequential(
    (0): Convolution2D(in_channels=3, out_channels=64, ksize=3, stride=1, pad=1, nobias=False, dilate=1, groups=1),
    (1): Convolution2D(in_channels=64, out_channels=64, ksize=3, stride=1, pad=1, nobias=False, dilate=1, groups=1),
    (2): lambda x: chainer.functions.max_pooling_2d(x, ksize=3, stride=1, pad=1),
  ),
  (seq2): Sequential(
    (0): Convolution2D(in_channels=3, out_channels=64, ksize=3, stride=1, pad=1, nobias=False, dilate=1, groups=1),
    (1): Convolution2D(in_channels=64, out_channels=64, ksize=3, stride=1, pad=1, nobias=False, dilate=1, groups=1),
    (2): functools.partial(<function max_pooling_2d at 0x102799b70>, ksize=3, stride=1, pad=1),
  ),
)
```

wkentaro commented:
I also added tests for repr of Link, Chain, ChainList and Sequential.

wkentaro commented:
I also added tests for repr of `Link`, `Chain`, `ChainList`, and `Sequential`.

Since the `__str__` of `Sequential` was implemented by you in https://github.com/chainer/chainer/pull/2918, could you give me your comment @mitmul ?
I implemented the `__repr__` of `Sequential` to be similar to that of `ChainList`, so it looks like https://github.com/chainer/chainer/pull/4853#issuecomment-396450174.

kmaehashi commented:
@wkentaro Sorry for the late response 🙇
We discussed this PR in the team:

>> Could you use `__str__` instead of `__repr__`?

> Do you have an idea of which of them we use?
> I referred this link: https://stackoverflow.com/questions/1436703/difference-between-str-and-repr for the general idea.

We think it's better to use  ``__str__`` instead of ``__repr__`` for this feature, because of two reasons:

* As the comment in the SO you pasted says, ``__repr__`` should be used for unambiguous expression. The current output does not include necessary information to reconstruct the Link (e.g., parameters are not included in the representation). (Please also see the description of ``__str__`` and ``__repr__`` in Python's offical docs: ``https://docs.python.org/3/reference/datamodel.html#object.__str__``)
* The output is too long for `__repr__`, which is used for REPL outputs. For example, users in interactive environment (ipython/notebook, etc.) may see long output when s/he forgot to assign Link/Chain to the variable; which is not a good experience.

>> I think it is better to expose `self._kwargs` (or function to register such keyword arguments) for so that user-defined Links can use it.

> Could you give me an example of that kind of use cases?

This is to expose this feature as an public API also to user-defined Links/Chains.
We think it is better to introduce an overridable method to `Link` like this:

```py
def kwargs(self):
    """Returns all keyword arguments as list of (key, value) tuple.

    The output is used by the string representation (:meth:`__str__`) of the Link.
    Links inheriting :class:`chainer.Link` can optionally override this method and
    provide better string representation.
    See the implementation of :meth:`chainer.links.Convolution2D.kwargs` for an example.
    """
    return []
```

Then in override this method in each Link.
The reason why we use an overridable method rather than an attribute is that some parameters are initialized after it is called.

niboshi commented:
I have some comments.
* Is there any better name than `kwargs`? It doesn't have to be specified as `kwargs` in the ctor. Something like `printable_params`?
* How about returning a generator instead of a list?

```
def printable_params(self):
    yield 'myparam', self.myparam
    for name, param in super(MyLink, self).printable_params():
        yield name, param
```

There is a choice whether parameters in a subclass should precede the base class.
I think it's better to precede, to make the specific parameters visible at first.
(but the other choice also makes sense.)

wkentaro commented:
@kmaehashi @niboshi I updated the code according to the comments. Could you review it?

kmaehashi commented on chainer/sequential.py:
```diff
@@ -229,40 +227,24 @@ def __reduce__(self):
+        reps = []
+        for index, layer in enumerate(self):
+            # Explore better representation by if-block.
+            if getattr(layer, '__name__', None) == '<lambda>':
```
* Could you add test case for this?
* It lacks support for `functools.partial` and `function.Function`?

wkentaro replied:
> Could you add test case for this?

I added the test.

wkentaro commented on chainer/link.py:
```diff
@@ -142,6 +142,22 @@ def __init__(self, **params):
+        )
+
+    @property
+    def hyper_params(self):
```
Can we use plural form, specifically `printable_specs`.

wkentaro commented on chainer/link.py:
```diff
@@ -791,6 +807,23 @@ def __init__(self, **links):
 
+    def __str__(self):
+        reps = []
+        for child in sorted(self.children(), key=lambda x: x.name):
```
Maybe `Link` and `Link` are not comparable?
```
% python model_chainlist.py
Traceback (most recent call last):
  File "model_chainlist.py", line 18, in <module>
    print(model)
  File "/Users/wkentaro/Documents/wkentaro/chainer/chainer/link.py", line 1008, in __str__
    index=index, rep=str(child),
  File "/Users/wkentaro/Documents/wkentaro/chainer/chainer/link.py", line 1008, in __str__
    index=index, rep=str(child),
  File "/Users/wkentaro/Documents/wkentaro/chainer/chainer/link.py", line 810, in __str__
    for child in sorted(self.children()):
TypeError: '<' not supported between instances of 'Linear' and 'Linear'
```

toslunar replied:
After #5119, the return of `self.children()` is already sorted by their names.  `x.name` is less reliable (than `self._children`) in some use cases (e.g. #2575).

wkentaro commented on chainer/link.py:
```diff
@@ -142,6 +142,22 @@ def __init__(self, **params):
+
+    @property
+    def hyper_params(self):
+        """Hyper parameters of this link.
```
I will update this with `printable_spec`.

mitmul commented:
@wkentaro Sorry for the delay. I didn't notice I've mentioned in this thread. This is a very nice feature. Could you resolve the conflicts? I want to check this PR after that and merge it.

wkentaro commented:
@mitmul Sorry for late. (I was occupied by the attendance to a conference.)
I rebased with resolving the conflicts and changed some code w.r.t the reviews by @kmaehashi:
- ed4cae4
- 5879a4d

Could you review this?

mitmul commented:
@wkentaro Thanks for resolving the conflicts. There seems to be a flake8 error. Could you also fix it? Thanks.

wkentaro commented:
Ah, sorry. I fixed it.

toslunar commented on chainer/links/connection/linear.py:
```diff
@@ -99,6 +99,7 @@ def __init__(self, in_size, out_size=None, nobias=False,
 
         if out_size is None:
             in_size, out_size = None, in_size
+        self.in_size = in_size
```
How about `self._in_size_spec`?  Leaving `self.in_size` `None` after `_initialize_params` is OK for printing, while it could be confusing for other purposes.

toslunar replied:
No.  `L.Linear` does not.  I guess there is a code where `L.Linear` is inherited and `self.in_size` and `_initialize_params` are patched to do so.

toslunar replied:
You are right.  I was just surprised at that an attribute is added for printing.  Comparing with other links, it's quite natural to have `Linear.in_size` attribute.

wkentaro commented:
I think I fixed according to the reviews. Could you specify what I need to do next?

mitmul commented:
@wkentaro Sorry for the late reply. Could you resolve the conflicts? After that, I think the remaining stuffs to do are just running Jenkins tests and merging.

@toslunar Could you handle this PR to merge?

toslunar commented:
> @toslunar Could you handle this PR to merge?

Sure.

wkentaro commented:
I fixed the conflicts.

toslunar commented on chainer/link.py:
```diff
@@ -198,6 +206,16 @@ def _device_id(self):
+        """Generator of printable specs of this link.
+
+        Usually, this indicates the arguments passed to the :meth:`__init__`.
+
```
Could you elaborate the docstring?  At least it could have
```
Yields:
    tuple of str and object: ...
```

toslunar commented on chainer/links/normalization/batch_normalization.py:
```diff
@@ -192,10 +192,20 @@ def __init__(self, size=None, decay=0.9, eps=2e-5, dtype=None,
+            ('use_gamma', use_gamma),
+            ('use_beta', use_beta),
+        ]
+
```
Could you revert unused changes?

wkentaro commented:
Thank you for the review. I updated the code, so could you check again?

toslunar commented:
Jenkins, test this please.

toslunar commented:
The above failure is due to a network error.

Jenkins, test this please.

toslunar commented:
Could you fix doctest(s)?  The first failure is
```
Document: reference/generated/chainer.Sequential
------------------------------------------------

Warning, treated as error:
**********************************************************************
File "reference/generated/chainer.Sequential.rst", line ?, in default
Failed example:
    print(model_C)  # doctest: +NORMALIZE_WHITESPACE
Expected:
    0       Linear  W(10, 10)       b(10,)
    1       relu
    2       Linear  W(10, 10)       b(10,)
    3       sigmoid
Got:
    Sequential(
      (0): Linear(in_size=10, out_size=10, nobias=False),
      (1): <function relu at 0x7faf8ba38950>,
      (2): Linear(in_size=10, out_size=10, nobias=False),
      (3): <function sigmoid at 0x7faf8ba3e8c8>,
    )

Makefile:176: recipe for target 'doctest' failed
```

wkentaro commented:
I fixed it in ecdb609.

toslunar commented on chainer/sequential.py:
```diff
@@ -85,10 +83,12 @@ class Sequential(_link.ChainList):
-        3       sigmoid
+        Sequential(
+          (0): Linear(in_size=10, out_size=10, nobias=False),
+          (1): <function relu at 0x7f45b1d36510>,
```
`doctest: +ELLIPSIS` could be used here to make the test pass.  It seems `NORMALIZE_WHITESPACE` is no longer required.
```suggestion
          (1): <function relu at 0x...>,
```
https://docs.python.org/3/library/doctest.html#doctest.ELLIPSIS

toslunar commented on chainer/sequential.py:
```diff
@@ -85,10 +83,12 @@ class Sequential(_link.ChainList):
+          (0): Linear(in_size=10, out_size=10, nobias=False),
+          (1): <function relu at 0x7f45b1d36510>,
+          (2): Linear(in_size=10, out_size=10, nobias=False),
+          (3): <function sigmoid at 0x7f45b1d3f510>,
```
ditto

wkentaro commented:
Thanks! I fixed it.

toslunar commented:
Jenkins, test this please.

wkentaro commented:
Thanks!

On Wed, Mar 20, 2019 at 4:37 Toshiki Kataoka <notifications@github.com>
wrote:

> Merged #4853 <https://github.com/chainer/chainer/pull/4853> into master.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/chainer/chainer/pull/4853#event-2215672403>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEHFk0_uuqFXpht84ZGFw0Q9KVAwq0WZks5vYbsfgaJpZM4Ub7l_>
> .
>
-- 
Kentaro Wada / 和田 健太郎
https://wkentaro.com <http://wkentaro.com>
